sdk/core/System.ClientModel/gen/BuildableAttributeUsageAnalyzer.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace System.ClientModel.SourceGeneration;
+
+/// <summary>
+/// Analyzer that ensures ModelReaderWriterBuildableAttribute is only applied to classes inheriting from ModelReaderWriterContext.
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class BuildableAttributeUsageAnalyzer : DiagnosticAnalyzer
+{
+    /// <inheritdoc/>
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+        ImmutableArray.Create(ModelReaderWriterContextGenerator.DiagnosticDescriptors.BuildableAttributeRequiresContext);
+
+    /// <inheritdoc/>
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context)
+    {
+        var namedType = (INamedTypeSymbol)context.Symbol;
+        if (namedType.TypeKind != TypeKind.Class)
+            return;
+
+        // Look for ModelReaderWriterBuildableAttribute
+        var hasBuildableAttribute = namedType.GetAttributes().Any(attr =>
+            attr.AttributeClass?.ToDisplayString() == "System.ClientModel.Primitives.ModelReaderWriterBuildableAttribute");
```

Comment:
Try to avoid matching by `ToDisplayString()` if you can, to avoid realizing strings when unnecessary. You could fetch this attribute type in a CompilationStart action and then use `SymbolEqualityComparer` to look for it.